File: macos/Sources/FloatingLyricCore/Support/AppError.swift
import Foundation

public enum AppError: Error, Equatable {
    case notConfigured
    case notLoggedIn
    case sessionExpired
    case network(String)
    case rateLimited(retryAfter: TimeInterval)
    case badResponse(status: Int)
    case portUnavailable
    case authCancelled
    case authUnavailable
    case authStateMismatch
    case premiumRequired
    case noActiveDevice
    case controlScopeMissing

    public var displayMessage: String {
        switch self {
        case .notConfigured:   return "Set up Spotify to begin"
        case .notLoggedIn:     return "Log in to Spotify"
        case .sessionExpired:  return "Session expired — log in again"
        case .network:         return "Offline"
        case .rateLimited:     return "Spotify is rate limiting — retrying"
        case .badResponse:     return "Spotify returned an unexpected response"
        case .portUnavailable: return "Port 8888–8890 in use. Free one and retry."
        case .authCancelled:   return "Login cancelled"
        case .authUnavailable: return "Could not open the Spotify login window"
        case .authStateMismatch:
            return "Login response did not match this request — try again"
        case .premiumRequired: return "Spotify Premium required to control playback"
        case .noActiveDevice:  return "No active Spotify device"
        case .controlScopeMissing:
            return "Log out and back in to enable playback controls"
        }
    }
}

File: windows/src/FloatingLyric.Core/AppError.cs
namespace FloatingLyric.Core;

/// <summary>Everything that can go wrong, and what to say about it.</summary>
public abstract record AppError
{
    public sealed record NotConfigured : AppError;
    public sealed record NotLoggedIn : AppError;
    public sealed record SessionExpired : AppError;
    public sealed record Network(string Detail) : AppError;
    public sealed record RateLimited(double RetryAfterSeconds) : AppError;
    public sealed record BadResponse(int Status) : AppError;
    public sealed record PortUnavailable : AppError;
    public sealed record AuthCancelled : AppError;
    public sealed record AuthStateMismatch : AppError;
    public sealed record PremiumRequired : AppError;
    public sealed record NoActiveDevice : AppError;
    public sealed record ControlScopeMissing : AppError;

    public string DisplayMessage => this switch
    {
        NotConfigured => "Set up Spotify to begin",
        NotLoggedIn => "Log in to Spotify",
        SessionExpired => "Session expired — log in again",
        Network => "Offline",
        RateLimited => "Spotify is rate limiting — retrying",
        BadResponse => "Spotify returned an unexpected response",
        PortUnavailable => "Port 8888–8890 in use. Free one and retry.",
        AuthCancelled => "Login cancelled",
        AuthStateMismatch => "Login response did not match this request — try again",
        PremiumRequired => "Spotify Premium required to control playback",
        NoActiveDevice => "No active Spotify device",
        ControlScopeMissing => "Log out and back in to enable playback controls",
        _ => "Something went wrong",
    };
}

public sealed class AppErrorException(AppError error) : Exception(error.DisplayMessage)
{
    public AppError Error { get; } = error;
}
